Categorifier/C/CExpr/Function.hs

```diff
@@ -92,7 +98,7 @@ generateTopLevelFunction' functionName rtg = do
         foldr
           (\inc acc -> acc <> Doc.line <> includeUserHeader inc)
           mempty
-          (Doc.pretty (functionName <> ".h") : funcallHeaders)
+          funcallHeaders
```

Comment:
I'd suggest making whether to include the header or not configurable. And I think the default option should be to include the header, because the no-header option is only useful for testing, right? Is there any non-test scenario where one wants to use `addForeignSource`?

Reply:
Made a configuration type `FunctionGenMode = Standard | AllInOne`. For ordinary usage, `Standard` is used, and for this TH usage that requires a single all-in-one source code, `AllInOne` is used.
-- 
I think in general, autointegration with TH is quite useful in production (see grad-descent example), not only for testing. 
There are a lot of opportunities to use inlined C code for numerical optimization because setting separate build steps for code generation and FFI integration is quite non-trivial (especially one wants to stay their project build in a simple cabal.project-like setup).

Reply:
> There are a lot of opportunities to use inlined C code for numerical optimization

That's true for using some C library, but does one ever need to generate C from Haskell, _and then_ embed the generated C in Haskell, unless they are testing codegen? I can't think of any cases. Anyway, this looks good to me.